src/shogun/neuralnets/NeuralMishLayer.cpp

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2014, Shogun Toolbox Foundation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its
+ * contributors may be used to endorse or promote products derived from this
+ * software without specific prior written permission.
+
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Written (W) 2020 Manjunath Bhat
+ */
+
+#include <shogun/neuralnets/NeuralMishLayer.h>
+#include <shogun/mathematics/Math.h>
+#include <shogun/lib/SGVector.h>
+
+using namespace shogun;
+
+NeuralMishLayer::NeuralMishLayer() : NeuralLinearLayer()
+{
+}
+
+NeuralMishLayer::NeuralMishLayer(int32_t num_neurons):
+NeuralLinearLayer(num_neurons)
+{
+}
+
+void NeuralMishLayer::compute_activations(SGVector<float64_t> parameters,
+    const std::vector<std::shared_ptr<NeuralLayer>>& layers)
+{
+	NeuralLinearLayer::compute_activations(parameters, layers);
+
+	int32_t len = m_num_neurons*m_batch_size;
+	for (int32_t i=0; i<len; i++)
```

Comment:
you could use a ranged loop here no?
```cpp
for (auto& activation: m_activations)
{
  activation *= std::tanh(std::log(1 + std::exp(activation)));
}

Reply:
or even better, std::transform with the same input and output iterators, that way it is clear that you are changing some values

Reply:
`std::log1p` instead of `std::log(1 + std::exp)` is better numerical accuracy but is still poor.

OpenCV CPU backend has nearly 2x faster and even more accurate [implementation](https://github.com/opencv/opencv/blob/4564b8a224545a54c545e9d2743cce3c0e1f1c5b/modules/dnn/src/layers/elementwise_layers.cpp#L760-L773).

Reply:
do you know what the accuracy tradeoff is?